### PR TITLE
Feature/test statistics

### PR DESCRIPTION
**Pull request by zelick**

### CODE REVIEW

File: Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionStatistic.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+
+            foreach (var exam in examsForQuestion)
+            {
+                var examQuestion = exam.ExamQuestions.FirstOrDefault(eq => eq.Question.Id == request.QuestionId);
+                if (examQuestion != null)
+                {
+                    totalAttempts++;
+
+                    if (IsCorrect(examQuestion))
```
Comment (Djoleoo) [inline Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionStatistic.cs:51]: use examQuestion.isCorrect()

File: Konteh/Konteh.BackOfficeApi/Features/Questions/QuestionController.cs [hunk 2]
```diff
@@ -1,11 +1,10 @@
 using Konteh.Domain;
 using MediatR;
-using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Mvc;
 
 namespace Konteh.BackOfficeApi.Features.Questions;
 
-[Authorize]
+
 [ApiController]
```
Comment (Djoleoo) [inline Konteh/Konteh.BackOfficeApi/Features/Questions/QuestionController.cs:8]: return [Authorize]

File: Konteh/back-office-app/src/app/features/questions/question-statistic/question-statistic.component.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import { Component, OnInit } from '@angular/core';
+import { QuestionClient } from '../../../api/api-reference';
+import { ActivatedRoute, Router } from '@angular/router';
+//import { CanvasJS } from '@canvasjs/angular-charts';
```
Comment (Djoleoo) [inline Konteh/back-office-app/src/app/features/questions/question-statistic/question-statistic.component.ts:4]: remove comments

File: Konteh/back-office-app/src/app/features/questions/question-category-statistics/question-category-statistics.component.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+         
+        response.forEach((element) => {
+          var dataPointElement = { label: element.categoryName, y: element.correctPercentage}
+          this.chartCategoryOptions.data[0].dataPoints.push(dataPointElement);
+        });
+        this.chartCategoryOptions = { ...this.chartCategoryOptions };
+        this.chartCategoryOptions.render(); 
+      },
+      error: (error) => {
+        console.error("Error getting question category statistics", error);
```
Comment (Djoleoo) [inline Konteh/back-office-app/src/app/features/questions/question-category-statistics/question-category-statistics.component.ts:52]: do you need console errors

**Review by Djoleoo — CHANGES_REQUESTED** [review]

File: Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionCategoryStatistic.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    public class Response
+    {
+        public string CategoryName { get; set; } = string.Empty;
+        public double CorrectPercentage { get; set; }
+    }
+
+    public class RequestHandler : IRequestHandler<Query, Response[]>
+    {
+        private readonly IRepository<Question> _questionRepository;
+        private readonly IRepository<ExamQuestion> _examQuestionRepository;
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionCategoryStatistic.cs:20]: what are all these repositories for?

File: Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionStatistic.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        private readonly IExamRepository _examRepository;
+        public RequestHandler(IRepository<Question> repository, IExamRepository examRepository)
+        {
+            _questionRepository = repository;
+            _examRepository = examRepository;
+        }
+
+        public async Task<Response> Handle(Query request, CancellationToken cancellationToken)
+        {
+            var question = _questionRepository.GetById(request.QuestionId).Result;
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionStatistic.cs:32]: Pulling the question just for this seems like overkill to me

File: Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionStatistic.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            if (question == null)
+            {
+                throw new NotFoundException();
+            }
+
+            var examsForQuestion = await _examRepository.GetByQuestion(request.QuestionId);
```
Comment (laslorajda) [inline Konteh/Konteh.BackOfficeApi/Features/Questions/GetQuestionStatistic.cs:38]: I don't know if pulling the exam is what we want here. I think examQuestions that have this questionId is what we want.
Then we can do totalAttempts = examQuestions.Count(x => x.IsCorrect())

File: Konteh/back-office-app/src/app/features/questions/question-category-statistics/question-category-statistics.component.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      next: (response: GetQuestionCategoryStatisticResponse[]) => {
+         
+        response.forEach((element) => {
+          var dataPointElement = { label: element.categoryName, y: element.correctPercentage}
+          this.chartCategoryOptions.data[0].dataPoints.push(dataPointElement);
+        });
+        this.chartCategoryOptions = { ...this.chartCategoryOptions };
+        this.chartCategoryOptions.render(); 
+      },
+      error: (error) => {}
```
Comment (laslorajda) [inline Konteh/back-office-app/src/app/features/questions/question-category-statistics/question-category-statistics.component.ts:51]: Empty error does nothing, remove it.

File: Konteh/back-office-app/src/app/features/questions/question-category-statistics/question-category-statistics.component.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    }
+    this.getQuestionCategoryStatistics();
+  }
+
+  getQuestionCategoryStatistics(){
+
+    this.questionClient.getCategoryQuestionStatistic().subscribe({
+      next: (response: GetQuestionCategoryStatisticResponse[]) => {
+         
+        response.forEach((element) => {
```
Comment (laslorajda) [inline Konteh/back-office-app/src/app/features/questions/question-category-statistics/question-category-statistics.component.ts:44]: maybe we can do something like this.chartCategoryOptions.data[0].dataPoints = response.map(e => doMappingHere)

File: Konteh/back-office-app/src/app/features/questions/question-form/question-form.component.html [hunk 10]
```diff
@@ -80,4 +80,6 @@ <h6 *ngIf="isEditingAnswer">EDIT ANSWER</h6>
       </div>
     </div>
   </form>
+
```
Comment (laslorajda) [inline Konteh/back-office-app/src/app/features/questions/question-form/question-form.component.html:83]: what 😄

File: Konteh/back-office-app/src/app/features/questions/question-form/question-form.component.ts [hunk 11]
```diff
@@ -194,3 +195,6 @@ export class QuestionFormComponent implements OnInit {
   }
 
 }
+
+
```
Comment (laslorajda) [inline Konteh/back-office-app/src/app/features/questions/question-form/question-form.component.ts:199]: huh

**Review by MilosNesh — APPROVED** [review]

**Review by Djoleoo — APPROVED** [review]

**Pull request merged**